File: apps/web/src/lib/api/onboarding.ts
import { apiClient } from './client';

export interface OnboardingStatusResponse {
    is_complete: boolean;
    step: string | null;
    completeness_score: number;
    has_extraction: boolean;
    extracted_sources: string[];
    profile_id?: string;
    writing_samples_count?: number;
}

export interface ExtractionSummary {
    source: string;
    fields_extracted: string[];
    highlights: Record<string, any>;
}

export interface ExtractResponse {
    success: boolean;
    summary: string;
    data?: Record<string, any>;
    extraction_summary?: ExtractionSummary;
    error?: string;
}

// Step-specific data interfaces
export interface ProfessionalStepData {
    current_role: string;
    industry: string;
    years_experience?: string;
    expertise_areas: string[];
    career_highlight?: string;
}

export interface InterestsStepData {
    interests: string[];
    aspirations?: string;
    topics_of_interest: string[];
}

export interface VoiceStepData {
    selected_post_ids: string[];
}

export interface StepSaveRequest {
    step_name: 'professional' | 'interests' | 'voice';
    professional_data?: ProfessionalStepData;
    interests_data?: InterestsStepData;
    voice_data?: VoiceStepData;
}

export interface StepSaveResponse {
    success: boolean;
    next_step?: string;
    message?: string;
}

// Conversational onboarding interfaces
export interface ChatMessage {
    role: 'user' | 'assistant';
    content: string;
}

export interface ConversationalExtractedData {
    current_role: string;
    industry: string;
    years_experience: string;
    expertise_areas: string[];
    career_highlight: string;
    interests: string[];
    topics_of_interest: string[];
    aspirations: string;
    tone_formal_casual: number | null;
    tone_technical_simple: number | null;
    tone_serious_playful: number | null;
    tone_humble_confident: number | null;
    bio_summary: string;
}

export interface ChatRequest {
    message: string;
    conversation_history: ChatMessage[];
    mode?: 'onboarding' | 'refinement';
}

export interface ChatResponse {
    response: string;
    extracted_data: ConversationalExtractedData;
    is_complete: boolean;
    conversation_history: ChatMessage[];
}

export interface LinkedInURLRequest {
    linkedin_url: string;
}

export interface LinkedInURLResponse {
    success: boolean;
    message: string;
    profile_extracted: boolean;
    posts_scraping_queued: boolean;
    career_timeline_extracted: boolean;
    suggested_topics: string[];
}

export interface ContentFocusRequest {
    primary_topics: string[];
    custom_topics: string[];
}

export interface ContentFocusResponse {
    success: boolean;
    topics_saved: number;
}

export const onboardingApi = {
    /**
     * Get onboarding status without modifying state (read-only)
     */
    getStatus: async (): Promise<OnboardingStatusResponse> => {
        const { data } = await apiClient.get<OnboardingStatusResponse>('/onboarding/status');
        return data;
    },

    /**
     * Submit LinkedIn profile URL for scraping profile + posts
     */
    submitLinkedInUrl: async (linkedinUrl: string): Promise<LinkedInURLResponse> => {
        const { data } = await apiClient.post<LinkedInURLResponse>('/onboarding/linkedin-url', {
            linkedin_url: linkedinUrl,
        });
        return data;
    },

    /**
     * Upload and extract data from LinkedIn profile PDF
     */
    uploadResume: async (file: File): Promise<ExtractResponse> => {
        const formData = new FormData();
        formData.append('file', file);
        const { data } = await apiClient.post<ExtractResponse>('/onboarding/upload-resume', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return data;
    },

    /**
     * Save step data during onboarding
     */
    saveStep: async (request: StepSaveRequest): Promise<StepSaveResponse> => {
        const { data } = await apiClient.post<StepSaveResponse>('/onboarding/save-step', request);
        return data;
    },

    /**
     * Complete onboarding and finalize profile
     */
    complete: async (): Promise<{ success: boolean; redirect_url: string }> => {
        const { data } = await apiClient.post<{ success: boolean; redirect_url: string }>('/onboarding/complete', {});
        return data;
    },

    /**
     * Save content focus topics during onboarding
     */
    saveContentFocus: async (primaryTopics: string[], customTopics: string[] = []): Promise<ContentFocusResponse> => {
        const { data } = await apiClient.post<ContentFocusResponse>('/onboarding/content-focus', {
            primary_topics: primaryTopics,
            custom_topics: customTopics,
        });
        return data;
    },

    /**
     * Conversational onboarding chat with Pixo
     */
    chat: async (request: ChatRequest): Promise<ChatResponse> => {
        const { data } = await apiClient.post<ChatResponse>('/onboarding/chat', request);
        return data;
    }
};

File: apps/api/app/schemas/onboarding.py
"""Onboarding request and response schemas."""
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class ExtractionSummary(BaseModel):
    """Summary of extracted data for UI display."""

    source: str
    fields_extracted: List[str]
    highlights: Dict[str, Any]


class OnboardingExtractResponse(BaseModel):
    """Response from extraction."""

    success: bool
    summary: str
    data: Optional[Dict[str, Any]] = None
    extraction_summary: Optional[ExtractionSummary] = None
    error: Optional[str] = None


class OnboardingCompleteResponse(BaseModel):
    """Response from completing onboarding."""

    success: bool
    redirect_url: str


class OnboardingStatusResponse(BaseModel):
    """Response for onboarding status check (read-only)."""

    is_complete: bool
    step: Optional[str] = None
    completeness_score: int = 0
    has_extraction: bool = False
    extracted_sources: List[str] = []


# Step-specific data schemas
class ProfessionalStepData(BaseModel):
    """Data from professional background step."""

    current_role: str
    industry: str
    years_experience: Optional[str] = None
    expertise_areas: List[str] = []
    career_highlight: Optional[str] = None


class InterestsStepData(BaseModel):
    """Data from interests/personality step."""

    interests: List[str] = []
    aspirations: Optional[str] = None
    topics_of_interest: List[str] = []


class VoiceStepData(BaseModel):
    """Data from voice selection step."""

    selected_post_ids: List[str] = []


class OnboardingStepSaveRequest(BaseModel):
    """Request to save onboarding step data."""

    step_name: str  # "professional", "interests", "voice"
    professional_data: Optional[ProfessionalStepData] = None
    interests_data: Optional[InterestsStepData] = None
    voice_data: Optional[VoiceStepData] = None


class OnboardingStepSaveResponse(BaseModel):
    """Response from saving step data."""

    success: bool
    next_step: Optional[str] = None
    message: Optional[str] = None


# ============================================================================
# Conversational Onboarding Schemas
# ============================================================================


class ChatMessage(BaseModel):
    """A single message in the conversation."""

    role: Literal["user", "assistant"]
    content: str


class ConversationalExtractedData(BaseModel):
    """Structured data extracted from the conversation."""

    # Professional
    current_role: str = ""
    industry: str = ""
    years_experience: str = ""
    expertise_areas: List[str] = Field(default_factory=list)
    career_highlight: str = ""

    # Interests
    interests: List[str] = Field(default_factory=list)
    topics_of_interest: List[str] = Field(default_factory=list)
    aspirations: str = ""

    # Tone sliders (0.0-1.0, None if not yet discussed)
    tone_formal_casual: Optional[float] = None
    tone_technical_simple: Optional[float] = None
    tone_serious_playful: Optional[float] = None
    tone_humble_confident: Optional[float] = None

    # Summary
    bio_summary: str = ""


class OnboardingChatRequest(BaseModel):
    """Request for conversational onboarding."""

    message: str
    conversation_history: List[ChatMessage] = Field(default_factory=list)


class OnboardingChatResponse(BaseModel):
    """Response from conversational onboarding."""

    response: str  # Pixo's reply
    extracted_data: ConversationalExtractedData  # Structured data from conversation
    is_complete: bool  # Has enough info to complete onboarding
    conversation_history: List[ChatMessage]  # Updated conversation history
